src/security-council-mgmt/SecurityCouncilMemberSyncAction.sol

```diff
@@ -3,22 +3,40 @@ pragma solidity 0.8.16;
 
 import "./interfaces/IGnosisSafe.sol";
 import "./SecurityCouncilMgmtUtils.sol";
+import "../UpgradeActionStorage.sol";
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 
 /// @notice Action contract for updating security council members. Used by the security council management system.
 ///         Expected to be delegate called into by an Upgrade Executor
-contract SecurityCouncilMemberSyncAction {
+contract SecurityCouncilMemberSyncAction is UpgradeActionStorage {
     error PreviousOwnerNotFound(address targetOwner, address securityCouncil);
     error ExecFromModuleError(bytes data, address securityCouncil);
+    error PrevActionNotExecuted(address securityCouncil, uint256 nonce);
 
     /// @dev Used in the gnosis safe as the first entry in their ownership linked list
     address public constant SENTINEL_OWNERS = address(0x1);
 
+    constructor(KeyValueStore _store)
+        UpgradeActionStorage(_store, "SecurityCouncilMemberSyncAction")
+    {}
+
     /// @notice Updates members of security council multisig to match provided array
     /// @dev    This function contains O(n^2) operations, so doesnt scale for large numbers of members. Expected count is 12, which is acceptable.
     /// Gnosis OwnerManager handles reverting if address(0) is passed to remove/add owner
     /// @param _securityCouncil The security council to update
     /// @param _updatedMembers  The new list of members. The Security Council will be updated to have this exact list of members
-    function perform(address _securityCouncil, address[] memory _updatedMembers) external {
+    function perform(address _securityCouncil, address[] memory _updatedMembers, uint256 _nonce)
+        external
+    {
+        // make sure the previous action was executed
+        if (_nonce > 0 && !_getExecuted(actionId(_securityCouncil, _nonce - 1))) {
```

Comment:
Instead of checking "true"/"false" we could instead store the nonce, and allow execution if the new nonce is higher. This would allow updates to fail - which actually we're ok with since each execution of the member sync action sets all the members to the latest state (it doesnt do a diff). We dont need all actions to execute, we just need it to be that an earlier update doesnt execute after a later one.

One advantage this might have is that retryable tickets can expire, if this were to happen with the current model then the process would halt and the sec council would need to unbrick it.

Reply:
Ahh very good point